Implement single expression QA checks

dodumosu opened the pull request:
This pull request adds some enhancements to QA expressions in order to fully resolve nditech/apollo-issues#39

dodumosu committed:
feat: add TRUE and FALSE constants to QA grammar

dodumosu committed:
feat: add support for single-line QA expressions

dodumosu committed:
feat: error out for QA on multivalue questions

dodumosu committed:
fix: add missing import

dodumosu committed:
fix: fix broken "missing" QA logic

this commit fixes an issue with the QA dashboard,
where checklists that should be marked as "missing"
are marked as "flagged"

dodumosu committed:
fix: broken logic in QA dashboard

dodumosu committed:
feat: add typing to QA processing

dodumosu committed:
fix: revert lookups to numeric type

dodumosu committed:
feat: add rendering of QA for
single-expression type

dodumosu committed:
feat: add data migration for QA checks

this commit adds a data migration for QA checks to single expressions

takinbo commented on migrations/versions/1e17be41a449_migrate_qa_expressions.py:
```diff
@@ -0,0 +1,40 @@
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    fetch_query = sa.sql.text(
+        "SELECT id, quality_checks FROM form WHERE quality_checks IS NULL OR quality_checks = 'null'::jsonb;")  # noqa
```
Can you review this SQL query? Shouldn't it be `IS NOT NULL`?

takinbo commented on migrations/versions/1e17be41a449_migrate_qa_expressions.py:
```diff
@@ -0,0 +1,40 @@
+    for form_id, quality_checks in op.execute(fetch_query).fetchall():
+        for quality_check in quality_checks:
+            if 'expression' not in quality_check:
+                quality_check['expression'] = qb.build_expression(
```
Since we are practically doing away with `criteria` for `quality_check` shouldn't we also clean up the code and not have portions of the code that will never be used going forward? That should be the purpose of the data migration.

takinbo commented on apollo/templates/admin/quality_assurance.html:
```diff
@@ -39,9 +39,13 @@
+              {%- if 'expression' in control -%}
+                <h5><span class="badge badge-pill badge-primary">{{ control.expression }}</span></h5>
+              {%- else -%}
               {%- for criterion in control.criteria %}
```
We would not be needing `criteria` anymore so the code should be cleaned up some.

takinbo commented on apollo/submissions/qa/query_builder.py:
```diff
@@ -350,11 +468,13 @@ def get_inline_qa_status(submission, condition):
-    if 'criteria' in logical_check:
+    if 'expression' in logical_check:
+        control_expression = logical_check.get('expression')
+    elif 'criteria' in logical_check:
```
No need for `criteria`

dodumosu replied:
the migration working depends on this staying unchanged.

takinbo replied:
What if this was just put in the migration itself?

dodumosu replied:
i'm also using it in the form import, which i have since updated

takinbo commented on apollo/formsframework/views_forms.py:
```diff
@@ -229,7 +229,9 @@ def quality_controls(view, form_id):
-            if 'criteria' in quality_check:
+            if 'expression' in quality_check:
+                quality_control['expression'] = quality_check['expression']
+            elif 'criteria' in quality_check:
```
Remove check for `criteria`

takinbo commented on apollo/submissions/qa/query_builder.py:
```diff
@@ -27,8 +29,10 @@
 name = r'[a-zA-Z_][a-zA-Z0-9_]*'
 lookup = "$" ("location" / "participant" / "submission") ("." / "@") name
 null = "NULL"
+true = "TRUE"
```
I don't think we have comparisons for `TRUE` or `FALSE`, it was put there as a hack when we were using the three-part criterion for the quality assurance and so isn't needed.

takinbo reviewed (changes requested):
We should have test cases for this change.

dodumosu committed:
fix: fix broken query in migration

dodumosu committed:
chore: make PEP8 compliant

dodumosu committed:
chore: clean up QA template

dodumosu committed:
chore: clean up QA view

dodumosu committed:
feat: update form exports for single QA

dodumosu committed:
feat: update form export to single expression QA

dodumosu committed:
fix: update form import/export

dodumosu committed:
chore: remove migrated QA criteria

dodumosu committed:
chore: remove Boolean constants

dodumosu committed:
fix: update migration script

- this commit fixes an issue with the migration script
where the user receives an `AttributeError` while running it
- it also doesn't assume that QA is present on the form

dodumosu committed:
chore: reorder migrations

dodumosu commented:
At this point, a rewrite would be better than a rebase.

dodumosu commented:
reopening to note, but still needs a lot of work
